### PR TITLE
DBOS Transact

### PR DESCRIPTION
**Pull request by kraftp**

### CODE REVIEW

File: docs/index.md [hunk 1]
```diff
@@ -2,35 +2,33 @@
...
-### What is DBOS?
+You want to build your next database-backed application with DBOS Transact because:
 
-DBOS is a transactional serverless framework and platform that helps you develop and deploy database-backed applications.
-You develop your applications in TypeScript and PostgreSQL with our [open-source SDK](https://github.com/dbos-inc/dbos-ts) then deploy them in minutes to DBOS Cloud.
-
-You want to build your next database-backed application with DBOS because:
-
-- **It's simple**.  Write your business logic using serverless functions, run them locally, and deploy them to the cloud in minutes.  Store all your data in Postgres&#8212;we'll manage the connections and transactions for you.
+- **It's simple**.  Write your business logic with serverless functions and either [self-host them](./tutorials/self-hosting.md) or deploy them to the cloud in minutes.  Store all your data in Postgres&#8212;we'll manage the connections and transactions for you.
```
Comment (apoliakov) [inline docs/index.md:12]: You don't want to put a second link over "deploy them to the cloud" -> cloud quickstart? I thought that might look good

File: docs/tutorials/self-hosting.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
+---
+sidebar_position: 25
+title: Self-Hosting
+description: Learn how to self-host DBOS Transact applications
+---
+
+You can run DBOS applications anywhere as long as they have a Postgres database to connect to.
```
Comment (qianl15) [inline docs/tutorials/self-hosting.md:7]:
```suggestion
You can run DBOS applications anywhere with `npx dbos start` as long as they have a Postgres database to connect to.
```

File: docs/api-reference/contexts.md [hunk 3]
```diff
@@ -106,6 +107,16 @@ readonly logger: DBOSLogger
 A reference to DBOS's logger.
 Please see our [logging tutorial](../tutorials/logging.md) for more information.
 
+#### `ctxt.span`
+
+```typescript
+readonly span: Span
+```
+
+An [OpenTelemetry Span](https://opentelemetry.io/docs/concepts/signals/traces/#spans) associated with this function.
```
Comment (maxdml) [inline docs/api-reference/contexts.md:116]:
```suggestion
An [OpenTelemetry Span](https://opentelemetry.io/docs/concepts/signals/traces/#spans) associated with this context.
```

File: docs/api-reference/contexts.md [hunk 4]
```diff
@@ -106,6 +107,16 @@ readonly logger: DBOSLogger
...
+You can assign custom trace attributes to this span.
+Please see our [self-hosting tutorial](../tutorials/self-hosting.md#configuring-otlp-telemetry) for more information.
```
Comment (maxdml) [inline docs/api-reference/contexts.md:118]: I would not link to this page here. I would rather link to the [span API](https://github.com/open-telemetry/opentelemetry-js/blob/f6a075be3204befa970cd4b7d6101419ee5d562e/api/src/trace/span.ts#L32).

File: docs/api-reference/static-analysis.md [hunk 5]
```diff
@@ -15,9 +15,8 @@ While the list of "gotchas" is long and easily neglected, the good news is that
 DBOS recommends using static analysis as an ingredient in a comprehensive security strategy.  As adding rule enforcement to a large, established codebase can be a hassle, DBOS recommends using tools from the beginning of a project, and therefore includes tool configuration in its [demo applications](https://github.com/dbos-inc/dbos-demo-apps) and [quickstart templates](../getting-started/quickstart.md).
 
 DBOS uses several techniques to ensure that static analysis is as productive as possible, with minimal hassle:
-* DBOS builds on popular frameworks, thereby leveraging community best-practices and tools integration.
+* DBOS Transact builds on popular frameworks, thereby leveraging community best-practices and tools integration.
 * DBOS focuses on analysis rules that detect incorrect API usage and potential security vulnerabilities, rather than nitpicking on coding style.
```
Comment (maxdml) [inline docs/api-reference/static-analysis.md:19]: Should we also name this DBOS Transact?

File: docs/index.md [hunk 6]
```diff
@@ -2,51 +2,42 @@
 sidebar_position: 1
 ---
 
-# Introduction
+# Welcome to DBOS!
 
-Welcome to DBOS!
+With these DBOS offerings, you can easily build scalable, reliable, and secure TypeScript applications:
```
Comment (maxdml) [inline docs/index.md:7]: I wouldn't say "offering" here. DBOS transact is and will always be open source. I think `With these DBOS solutions` would represent more accurately our commitment to OSS.

File: docs/index.md [hunk 7]
```diff
@@ -2,51 +2,42 @@
...
+- **DBOS Cloud** - A transactional serverless platform for deploying DBOS Transact applications to the cloud with built-in high performance, scalability, fault tolerance, and database time travel.
 
 You want to build your next database-backed application with DBOS because:
 
-- **It's simple**.  Write your business logic using serverless functions, run them locally, and deploy them to the cloud in minutes.  Store all your data in Postgres&#8212;we'll manage the connections and transactions for you.
 - **It's reliable by default**.  If your workflows are interrupted for any reason, they [will always resume from where they left off](./tutorials/workflow-tutorial#reliability-guarantees).  Reliable message delivery is [built in](./tutorials/workflow-communication-tutorial#reliability-guarantees-1). Idempotency is [built in](./tutorials/idempotency-tutorial).
+- **It's simple**.  Write your business logic with serverless functions and either [self-host them](./tutorials/self-hosting.md) or [deploy them to the cloud](./getting-started/quickstart.md#deploying-to-dbos-cloud) in minutes. Store your data in any Postgres&mdash;we'll manage the transactions for you.
```
Comment (maxdml) [inline docs/index.md:16]:
```suggestion
- **It's simple**.  Write your business logic with serverless functions and either [self-host them](./tutorials/self-hosting.md) or [deploy them to DBOS cloud](./getting-started/quickstart.md#deploying-to-dbos-cloud) in minutes. Store your data in any Postgres&mdash;we'll manage the transactions for you.
```

Comment (maxdml) [reply docs/index.md:16]: manage the transactions -> manage the state?

File: docs/tutorials/self-hosting.md [hunk 8]
```diff
@@ -0,0 +1,98 @@
+---
+sidebar_position: 25
+title: Self-Hosting
+description: Learn how to self-host DBOS Transact applications
+---
+
+You can run DBOS applications anywhere with `npx dbos start` as long as they have a Postgres database to connect to.
+This guide describes tools that make it easy to self-host DBOS applications in production.
```
Comment (maxdml) [inline docs/tutorials/self-hosting.md:8]:
```suggestion
This guide describes tools you can use in your hosting environment to make the most of DBOS transact.
```
These tools don't make it easy to host DBOS apps, but rather allows one to hook with all the cool stuff DBOS transact offer (namely: workflow recovery and built-in observability.)

File: docs/tutorials/self-hosting.md [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+## Admin API
+
+DBOS applications expose an admin API, fixed at one above the main DBOS application port (the main port defaults to port 3000, so the admin API defaults to port 3001).
```
Comment (maxdml) [inline docs/tutorials/self-hosting.md:12]: We should explicitly state the admin API is (main port + 1)

File: docs/tutorials/self-hosting.md [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+### Workflow Recovery
+
+
+- **Endpoint**: `/dbos-workflow-recovery`
+- **Method**: POST
+- **Description**: Recovers all pending workflows associated with input [executor IDs](#managing-workflow-recovery). Following our [reliability guarantees](./workflow-tutorial.md#reliability-guarantees), all workflows will resume from where they left off. Returns the UUIDs of all workflows recovered.
+- **Request Body Format**: JSON list of executors whose pending workflows to recover.
+  - **Example**:
+    ```json
+      ["executor-id-1", "executor-id-2", "..."]
```
Comment (maxdml) [inline docs/tutorials/self-hosting.md:32]:
```suggestion
      ["executor-id-1", "executor-id-2", "..."]
    ```
```

File: docs/tutorials/self-hosting.md [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+- **Response**:
+  - **Status Code**: 200 OK on successful recovery initiation; otherwise, appropriate error codes.
+  - **Body Format**: JSON list of UUIDs representing the workflows that were successfully queued for recovery.
+  - **Example**:
+    ```json
+    ["workflow-uuid-1", "workflow-uuid-2", "..."]
+    ```
```
Comment (maxdml) [inline docs/tutorials/self-hosting.md:39]: So what happens if only part of the workflows are successfully recovered? 500 and a list of successfully recovered workflows in the body?

File: docs/tutorials/self-hosting.md [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+
+## Managing Workflow Recovery
+
+By default, when a DBOS application starts up, it recovers all pending workflows, resuming them from where they left off following our [reliability guarantees](./workflow-tutorial.md#reliability-guarantees).
+This behavior works well when you're only running a single instance of an application, as it guarantees that every time the server is restarted, it resumes all workflows from where they left off.
+However, it is less ideal for a distributed setting where you're running many instances of an application on different servers.
+
+To manage recovery in a distributed setting, you can assign each instance of an application an executor ID by setting the `DBOS__VMID` environment variable.
+This causes the application instance to assign every workflow it executes to that executor ID.
```
Comment (maxdml) [inline docs/tutorials/self-hosting.md:69]:
```suggestion
This causes the application instance to associate every workflow it executes with that executor ID.
```

**Review by maxdml — APPROVED** [review]

**Pull request merged**